File: app/src/main/cpp/wn-steam-client/src/cmsg_protobuf_header.cpp
#include "wn_steam/cmsg_protobuf_header.h"

#include "wn_steam/proto_wire.h"

namespace wn_steam {

void CMsgProtoBufHeader::serialize(std::vector<uint8_t>& out) const {
    proto::Writer w(out);

    // 1 fixed64 steamid
    if (steamid != 0) {
        w.tag(1, proto::WireType::Fixed64);
        // fixed64 inline (no helper for arbitrary fixed64 here — Writer
        // exposes fixed64_field but that skips zero-by-default; we already
        // checked).
        for (int i = 0; i < 8; ++i) out.push_back(static_cast<uint8_t>(steamid >> (i * 8)));
    }

    // 2 int32 client_sessionid
    w.int32_field(2, client_sessionid);

    // 3 uint32 routing_appid
    w.uint32_field(3, routing_appid);

    // 10 fixed64 jobid_source — emit only when SET. The earlier comment
    // claiming "always emit" was wrong; SteamKit2's protobuf-net codegen
    // uses ShouldSerialize() which returns false when the field is null,
    // and JavaSteam's protobuf-java builder only emits fields the caller
    // explicitly set. The wire convention is "field absent = no job",
    // not "field present + sentinel value = no job".
    if (jobid_source != kInvalidJobId) {
        w.tag(10, proto::WireType::Fixed64);
        for (int i = 0; i < 8; ++i) out.push_back(static_cast<uint8_t>(jobid_source >> (i * 8)));
    }

    // 11 fixed64 jobid_target — same rule. CRITICAL: emitting this on an
    // outbound service-method call causes Steam to interpret the message
    // as a *reply* to a (nonexistent) job 0xFFFF..., which it silently
    // rejects with a stub response containing only `interval=5.0` and an
    // empty `extended_error_message`. This was the silent-fail bug for
    // BeginAuthSessionViaCredentials.
    if (jobid_target != kInvalidJobId) {
        w.tag(11, proto::WireType::Fixed64);
        for (int i = 0; i < 8; ++i) out.push_back(static_cast<uint8_t>(jobid_target >> (i * 8)));
    }

    // 12 string target_job_name
    w.string_field(12, target_job_name);

    // 14 int32 eresult — emit only on responses that explicitly set it.
    // Clients (us) normally never set eresult on outbound, so this branch
    // is a no-op in practice (default = -1, our sentinel for "unset").
    if (eresult >= 0) {
        w.tag(14, proto::WireType::Varint);
        w.varint(static_cast<uint64_t>(static_cast<int64_t>(eresult)));
    }

    // 15 string error_message
    w.string_field(15, error_message);

    // 29 uint32 realm
    w.uint32_field(29, realm);

    // 21 uint64 messageid
    w.uint64_field(21, messageid);

    // 34 uint64 token_id
    w.uint64_field(34, token_id);
}

std::optional<CMsgProtoBufHeader>
CMsgProtoBufHeader::deserialize(std::span<const uint8_t> bytes) noexcept {
    proto::Reader r(bytes);
    CMsgProtoBufHeader h;

    while (!r.eof()) {
        auto tag = r.next_tag();
        if (!tag) {
            if (!r.ok()) return std::nullopt;
            break;
        }
        switch (tag->field_number) {
            case 1:  // steamid (fixed64)
                if (tag->wire_type != proto::WireType::Fixed64) return std::nullopt;
                if (auto v = r.fixed64(); v) h.steamid = *v; else return std::nullopt;
                break;
            case 2:  // client_sessionid (int32 varint)
                if (auto v = r.i32(); v) h.client_sessionid = *v; else return std::nullopt;
                break;
            case 3:  // routing_appid (uint32 varint)
                if (auto v = r.u32(); v) h.routing_appid = *v; else return std::nullopt;
                break;
            case 10: // jobid_source (fixed64)
                if (tag->wire_type != proto::WireType::Fixed64) return std::nullopt;
                if (auto v = r.fixed64(); v) h.jobid_source = *v; else return std::nullopt;
                break;
            case 11: // jobid_target (fixed64)
                if (tag->wire_type != proto::WireType::Fixed64) return std::nullopt;
                if (auto v = r.fixed64(); v) h.jobid_target = *v; else return std::nullopt;
                break;
            case 12: // target_job_name (string)
                if (auto v = r.string(); v) h.target_job_name = std::move(*v); else return std::nullopt;
                break;
            case 14: // eresult (int32 varint)
                if (auto v = r.i32(); v) h.eresult = *v; else return std::nullopt;
                break;
            case 15: // error_message (string)
                if (auto v = r.string(); v) h.error_message = std::move(*v); else return std::nullopt;
                break;
            case 21: // messageid (uint64 varint)
                if (auto v = r.u64(); v) h.messageid = *v; else return std::nullopt;
                break;
            case 29: // realm (uint32 varint)
                if (auto v = r.u32(); v) h.realm = *v; else return std::nullopt;
                break;
            case 34: // token_id (uint64 varint)
                if (auto v = r.u64(); v) h.token_id = *v; else return std::nullopt;
                break;
            default:
                if (!r.skip(tag->wire_type)) return std::nullopt;
                break;
        }
    }
    return h;
}

}  // namespace wn_steam

File: app/src/main/cpp/wn-steam-client/rust/src/cmsg_protobuf_header.rs
use crate::proto_wire::{Reader, WireType, Writer};

pub const INVALID_JOB_ID: u64 = u64::MAX;

#[derive(Clone, Debug, Eq, PartialEq)]
pub struct CMsgProtoBufHeader {
    pub steamid: u64,
    pub client_sessionid: i32,
    pub routing_appid: u32,
    pub jobid_source: u64,
    pub jobid_target: u64,
    pub target_job_name: String,
    pub eresult: i32,
    pub error_message: String,
    pub realm: u32,
    pub messageid: u64,
    pub token_id: u64,
}

impl Default for CMsgProtoBufHeader {
    fn default() -> Self {
        Self {
            steamid: 0,
            client_sessionid: 0,
            routing_appid: 0,
            jobid_source: INVALID_JOB_ID,
            jobid_target: INVALID_JOB_ID,
            target_job_name: String::new(),
            eresult: -1,
            error_message: String::new(),
            realm: 0,
            messageid: 0,
            token_id: 0,
        }
    }
}

impl CMsgProtoBufHeader {
    pub fn serialize(&self, out: &mut Vec<u8>) {
        let mut writer = Writer::new(out);

        if self.steamid != 0 {
            writer.tag(1, WireType::Fixed64);
            writer.fixed64_field_force_body(self.steamid);
        }
        writer.int32_field(2, self.client_sessionid);
        writer.uint32_field(3, self.routing_appid);
        if self.jobid_source != INVALID_JOB_ID {
            writer.tag(10, WireType::Fixed64);
            writer.fixed64_field_force_body(self.jobid_source);
        }
        if self.jobid_target != INVALID_JOB_ID {
            writer.tag(11, WireType::Fixed64);
            writer.fixed64_field_force_body(self.jobid_target);
        }
        writer.string_field(12, &self.target_job_name);
        if self.eresult >= 0 {
            writer.tag(14, WireType::Varint);
            writer.varint(self.eresult as i64 as u64);
        }
        writer.string_field(15, &self.error_message);
        writer.uint32_field(29, self.realm);
        writer.uint64_field(21, self.messageid);
        writer.uint64_field(34, self.token_id);
    }

    pub fn deserialize(bytes: &[u8]) -> Option<Self> {
        let mut reader = Reader::new(bytes);
        let mut header = Self::default();

        while !reader.eof() {
            let Some(tag) = reader.next_tag() else {
                return reader.ok().then_some(header);
            };
            match tag.field_number {
                1 => {
                    if tag.wire_type != WireType::Fixed64 {
                        return None;
                    }
                    header.steamid = reader.fixed64()?;
                }
                2 => header.client_sessionid = reader.i32()?,
                3 => header.routing_appid = reader.u32()?,
                10 => {
                    if tag.wire_type != WireType::Fixed64 {
                        return None;
                    }
                    header.jobid_source = reader.fixed64()?;
                }
                11 => {
                    if tag.wire_type != WireType::Fixed64 {
                        return None;
                    }
                    header.jobid_target = reader.fixed64()?;
                }
                12 => header.target_job_name = reader.string()?,
                14 => header.eresult = reader.i32()?,
                15 => header.error_message = reader.string()?,
                21 => header.messageid = reader.u64()?,
                29 => header.realm = reader.u32()?,
                34 => header.token_id = reader.u64()?,
                _ => {
                    if !reader.skip(tag.wire_type) {
                        return None;
                    }
                }
            }
        }

        Some(header)
    }
}

trait WriterFixedBodies {
    fn fixed64_field_force_body(&mut self, v: u64);
}

impl WriterFixedBodies for Writer<'_> {
    fn fixed64_field_force_body(&mut self, v: u64) {
        self.raw_bytes(&v.to_le_bytes());
    }
}

#[cfg(test)]
mod tests {
    use super::*;

    #[test]
    fn serializes_only_set_job_ids() {
        let h = CMsgProtoBufHeader {
            steamid: 0x1122_3344_5566_7788,
            client_sessionid: 17,
            routing_appid: 480,
            target_job_name: "Authentication.BeginAuthSessionViaCredentials#1".to_string(),
            ..Default::default()
        };
        let mut bytes = Vec::new();
        h.serialize(&mut bytes);
        let parsed = CMsgProtoBufHeader::deserialize(&bytes).unwrap();
        assert_eq!(parsed, h);
        assert!(!bytes
            .windows(2)
            .any(|w| w == [0x51, 0xff] || w == [0x59, 0xff]));
    }

    #[test]
    fn rejects_wrong_fixed_wire_type() {
        let bytes = [0x08, 0x01];
        assert_eq!(CMsgProtoBufHeader::deserialize(&bytes), None);
    }
}
